File: app/src/main/java/com/innov8/memeit/Fragments/SignUpFragment.java
package com.innov8.memeit.Fragments;

import android.content.Intent;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;

import com.innov8.memeit.Activities.AuthActivity;
import com.innov8.memeit.CustomClasses.CustomMethods;
import com.innov8.memeit.R;
import com.memeit.backend.MemeItAuth;
import com.memeit.backend.dataclasses.User;
import com.memeit.backend.utilis.OnCompleteListener;

public class SignUpFragment extends AuthFragment implements View.OnClickListener {
    public SignUpFragment() {
    }

    EditText userNameV;
    EditText emailV;
    EditText passwordV;


    OnCompleteListener<Void> onCompleteListener;
    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_signup2, container, false);

        userNameV = view.findViewById(R.id.signup_username);
        emailV = view.findViewById(R.id.signup_email);
        passwordV = view.findViewById(R.id.signup_password);
        actionButton = view.findViewById(R.id.signup_btn);

        CustomMethods.makeEditTextsAvenir(getActivity(), view, R.id.signup_password, R.id.signup_username, R.id.signup_confirm);

        /* This sets the bottom padding that makes the views not go underneath the navigation bar */
        view.findViewById(R.id.rel).setPadding(
                0,
                (int) (16 * getResources().getDisplayMetrics().density + 0.5f),
                0,
                (int) (10 * getResources().getDisplayMetrics().density + 0.5f) + AuthActivity.getSoftButtonsBarHeight(getActivity()));
        view.findViewById(R.id.rel).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {

            }
        });
        actionButton.setOnClickListener(this);
        view.findViewById(R.id.signup_google).setOnClickListener(this);
        view.findViewById(R.id.signup_facebook).setOnClickListener(this);
        view.findViewById(R.id.rel).setOnClickListener(this);
        onCompleteListener = new OnCompleteListener<Void>() {
            @Override
            public void onSuccess(Void aVoid) {
                setLoading(false);
                AuthActivity a = ((AuthActivity) SignUpFragment.this.getActivity());
                a.setCurrentFragment(AuthActivity.FRAGMENT_SETUP);
            }

            @Override
            public void onFailure(Error error) {
                //todo if it is username conflict suggest a new one
                setLoading(false);
                getAuthActivity().showError(error.getMessage());
            }
        };
        return view;
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == MemeItAuth.GOOGLE_SIGNIN_REQUEST_CODE) {
            MemeItAuth.getInstance().handleGoogleSignUpResult(data, new OnCompleteListener<User>() {
                @Override
                public void onSuccess(User user) {
                   getAuthActivity().setupFragment.fromGoogle(user.getName(), user.getImageUrl());
                   getAuthActivity().setCurrentFragment(AuthActivity.FRAGMENT_SETUP);
                }
                @Override
                public void onFailure(Error error) {
                   getAuthActivity().showError(error.getMessage());
                }
            });
        }
    }

    @Override
    public void onClick(View view) {
        switch (view.getId()) {
            case R.id.signup_btn:
                signupWithUsername();
                break;
            case R.id.signup_google:
                MemeItAuth.getInstance().signInWithGoogle(getActivity());
                break;
            case R.id.signup_facebook:
                //todo facebook signup
                break;
            case R.id.rel:
                getAuthActivity().setCurrentFragment(AuthActivity.FRAGMENT_LOGIN);
                break;
        }
    }

    private void signupWithUsername() {
        String username = userNameV.getText().toString();
        String email = emailV.getText().toString();
        String password = passwordV.getText().toString();
        if (!CustomMethods.isUsernameValid(username)){
            getAuthActivity().showError("Username should at least be 5 in length!");
        }else if(!CustomMethods.isEmailValid(email)){//todo change back to 8
            getAuthActivity().showError("Invalid Email!");
        }else if(password.length() <= 1){//todo change back to 8
            getAuthActivity().showError("Password should at least be 8 in length!");
        }else{
            setLoading(true);
            MemeItAuth.getInstance().signUpWithUsername(username,email, password, onCompleteListener);
        }
    }
}

File: app/src/main/java/com/innov8/memeit/Fragments/SignUpFragment.kt
package com.innov8.memeit.Fragments

import android.content.Context.INPUT_METHOD_SERVICE
import android.content.Intent
import android.os.Bundle
import android.text.Editable
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.inputmethod.InputMethodManager
import androidx.constraintlayout.widget.ConstraintSet
import androidx.transition.TransitionManager
import com.facebook.CallbackManager
import com.innov8.memeit.Activities.AuthActivity
import com.memeit.backend.dataclasses.UsernameAuthRequest
import com.memeit.backend.MemeItClient
import kotlinx.android.synthetic.main.fragment_signup2.*
import com.innov8.memeit.R
import com.memeit.backend.dataclasses.FacebookInfo
import com.memeit.backend.dataclasses.GoogleInfo
import com.memeit.backend.dataclasses.SignInMethod

class SignUpFragment : AuthFragment() {
    lateinit var callbackManager: CallbackManager

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        callbackManager= CallbackManager.Factory.create()
    }
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_signup2, container, false)
    }
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        actionButton = signup_btn

        signup_confirm.setOnEditorActionListener { v, _, _ ->
            val mgr = context!!.getSystemService(INPUT_METHOD_SERVICE) as InputMethodManager
            mgr.hideSoftInputFromWindow(v.windowToken, 0)
            signupWithUsername()
            true
        }
        rel.setPadding(
                0,
                (16 * resources.displayMetrics.density + 0.5f).toInt(),
                0,
                (10 * resources.displayMetrics.density + 0.5f).toInt() + AuthActivity.getSoftButtonsBarHeight(activity))
        rel.setOnClickListener(View.OnClickListener { })
        actionButton.setOnClickListener { signupWithUsername() }
        signup_google.setOnClickListener {
            MemeItClient.Auth.requestGoogleInfo(authActivity)
        }
        signup_facebook.setOnClickListener {
            MemeItClient.Auth.requestFacebookInfo(authActivity,callbackManager,{
                setFacebookMode(it)
            },onError)

        }
        rel.setOnClickListener { authActivity.setCurrentFragment(AuthActivity.FRAGMENT_LOGIN) }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        callbackManager.onActivityResult(requestCode, resultCode, data)
        if (requestCode == MemeItClient.Auth.GOOGLE_SIGN_IN_REQUEST_CODE && data != null) {
            MemeItClient.Auth.extractGoogleInfo(data, {
                setGoogleMode(it)
            }, onError)
        }
    }


    var signInMethod=SignInMethod.USERNAME

    var ginfo: GoogleInfo? = null
    var fbinfo: FacebookInfo? = null
    private fun setAcceptUsernameMode() {
        val set2 = ConstraintSet()
        set2.clone(context, R.layout.fragment_signup2_username)
        TransitionManager.beginDelayedTransition(signup_contraint)
        set2.applyTo(signup_contraint)
        actionButton.text = "Continue"
    }
    private fun setGoogleMode(info:GoogleInfo){
        signInMethod=SignInMethod.GOOGLE
        this.ginfo = info
        setAcceptUsernameMode()
    }
    private fun setFacebookMode(info:FacebookInfo){
        signInMethod=SignInMethod.FACEBOOK
        this.fbinfo =info
        signup_username.text?.append("${info.firstName}${info.lastName}")
        setAcceptUsernameMode()
    }


    private fun signupWithUsername() {
        if (mLoading) return
        val username = signup_username.text.toString()
        when (signInMethod) {
            SignInMethod.GOOGLE -> {
                ginfo ?: return
                val greq = ginfo!!.toSignUpReq(username)
                MemeItClient.Auth.signUpWithGoogle(greq, {
                    authActivity.setupFragment.fromGoogle(ginfo!!.name, ginfo!!.imageUrl)
                    authActivity.setCurrentFragment(AuthActivity.FRAGMENT_SETUP)
                }, onError)
            }
            SignInMethod.FACEBOOK->{
                fbinfo ?: return
                val freq = fbinfo!!.toSignUpReq(username)
                MemeItClient.Auth.signUpWithFacebook(freq, {
                    val name="${fbinfo!!.firstName} ${fbinfo!!.lastName}"
                    authActivity.setupFragment.fromGoogle(name, fbinfo!!.imageUrl)
                    authActivity.setCurrentFragment(AuthActivity.FRAGMENT_SETUP)
                }, onError)
            }
            SignInMethod.USERNAME -> {
                val email = signup_email.text.toString()
                val password = signup_password.text.toString()
                val passwordConfrim = signup_confirm.text.toString()

                val req = UsernameAuthRequest(username, password, email)
                val errors = req.validate()
                when {
                    errors.isNotEmpty() -> authActivity.showError(errors.joinToString(", "))
                    password != passwordConfrim -> authActivity.showError("Password does not match!")
                    else -> {
                        setLoading(true)
                        MemeItClient.Auth.signUpWithUsername(req, {
                            setLoading(false)
                            authActivity.setCurrentFragment(AuthActivity.FRAGMENT_SETUP)
                        }, onError)
                    }
                }
            }
        }
    }
}
